### PR TITLE
feat(navigator): Implement a full mission route cache 

### PR DESCRIPTION
**Pull request by JonasPerolini**
### Route Cache

Normal mission execution only keeps a small sliding window of mission items in RAM, but route planning needs random, non-blocking access to the whole mission to project points against every segment.

`MissionRouteCache` provides this by composing an optional full-mission cache with the existing safe-point and mission-land caches.

```text
MissionRouteCache
|-- FullMissionCache              [0 ... CONFIG_NAVIGATOR_FULL_MISSION_CACHE_SIZE - 1]
|-- safe-point cache              [all uploaded safe-point Dataman items]
|-- mission-land item cache       [published mission land index]
`-- safe-point stats reader       [DM_KEY_SAFE_POINTS_STATE async state machine]
```

`FullMissionCache` owns the mission buffer, its Dataman client, generation and request state, retries, and load-time perf counter.

The full mission is stored as a flat `mission_item_s` array. It is loaded asynchronously one item at a time. While a read is pending, the Dataman response subscription is added to Navigator's poll list. A matching response advances the cache and queues the next item immediately, so loading is not limited by Navigator's normal update rate.

The response subscription is polled only while a read is pending because `dataman_response` is shared between clients. Polling it while idle could keep Navigator awake on unrelated responses. Dataman-only wakeups advance the cache immediately, but normal Navigator processing remains at its existing 20 Hz local-position speed, with a 50 ms fallback.

Missions larger than the configured capacity are reported once and remain uncached.

Route-provider reads are cache-only and never block Navigator on Dataman or the SD card. `getMissionView()` borrows the complete mission as a `const mission_item_s *` instead of copying it. The view includes the mission identity and cache generation so callers can detect whether it became stale with `missionViewStillValid()`.

A mission view is exposed only after every item has loaded successfully. The cache tracks the mission ID, item count, and Dataman bank and starts a new generation when the source changes. Responses belonging to an older generation are consumed but not published.

A failed read schedules a bounded exponential backoff and resumes from the failed index, so already loaded items are not read again.

Safe points have their own asynchronous state machine. They reload when their source identity changes and are not exposed while a replacement generation is still loading. Mission-land readiness is tracked independently.

### Cache Size

`CONFIG_NAVIGATOR_FULL_MISSION_CACHE_SIZE` sets the maximum number of mission items reserved for full-route planning. The default is 500 on POSIX/SITL and 0 otherwise.

A value of 0 compiles out the full-mission buffer and its Dataman client, so boards that enable a route-planning consumer must size it for their RAM budget. Each reserved item uses 56 bytes of heap allocated once when Navigator starts:

- 500 items: 28 kB
- 300 items: 16.8 kB

The safe-point and mission-land caches are independent of this option and remain available on every board.

The full-cache load time (including retry or delays) can be measured on hardware with:

```sh
perf reset
# Upload or change a non-empty mission
perf
```

### Cache Coherency

The same mission item can be represented in the full-mission buffer and, for the published landing item, in the dedicated mission-land cache.

After a successful active-mission Dataman write, `MissionRouteCache::syncMissionItem()` keeps these cached representations coherent. This currently happens when `MissionBase` increments or resets a `DO_JUMP` counter.

For an already-ready full cache, the updated item is patched directly into the mission array and the cache generation advances. The complete mission remains ready, but any previously borrowed `MissionView` becomes stale and must not be reused. Reloading the complete mission would unnecessarily make route data unavailable and issue one read per mission item after every counter update.

If the cache is still loading:

- An item in the already-loaded prefix is patched in place.
- A write to the item currently being read makes that response stale, so it is discarded and the item is read again.
- An unread item requires no immediate cache change because the ongoing load will read its updated value from Dataman.

The mission-land cache is synchronized independently. It is patched when possible and invalidated or reloaded when the published land item changes type or when an older read may still be pending.

`syncMissionItem()` is used in this PR for `DO_JUMP` counter coherency. The full-mission `MissionView` does not yet have a production route-planning consumer; that comes with the follow-up PR: https://github.com/PX4/PX4-Autopilot/pull/27687

### CODE REVIEW

**Comment by JonasPerolini** [comment]
Hey @bkueng, would you have time to review this PR please?

**Comment by JonasPerolini** [comment]
State machine summary:

`RTL::updateDatamanCache()` calls `MissionRouteCache::update(_mission_sub.get())`. 
This happens while RTL is inactive and active. The important path is inactive, before `setRtlTypeAndDestination()` is periodically recomputed.

On mission identity change, `updateMissionCache()` invalidates the RAM cache, queues all mission items with `DatamanCache::load()`, sets `validation_pending = true`.

`DatamanCache::update()` advances async dataman reads. The loaded items: `DatamanCache::_items[].response.data` are in RAM once each item reaches `ResponseReceived`.

Once loading stops, `missionCacheFullyLoaded()` validates every queued item using `loadWait(..., kCacheOnlyLoadWait)`, with timeout 0. At this point a miss means that the async preload did not work so it should retry, not block.

Only after that validation does `_mission.ready = true`.

Planner/provider reads call `loadMissionItem()` / `getMissionLandItem()` / `loadSafePointItem()`. These use timeout 0 because they are reading from an already validated RAM cache, not calling SD card reads.

File: src/modules/navigator/rtl.cpp [hunk 1]
```diff
@@ -501,44 +450,50 @@ void RTL::findRtlDestination(DestinationType &destination_type, PositionYawSetpo
...
-			}
+				const bool land_item_pending = mission_route_cache != nullptr
+							       && mission_route_cache->missionLandItemUpdatePending();
+
+				if (!land_item_pending) {
+					/* Not supposed to happen unless the datamanager can't access the SD card, etc. */
+					mavlink_log_critical(_navigator->get_mavlink_log_pub(), "Mission land item could not be read.\t");
+					events::send(events::ID("rtl_failed_to_read_land_item"), events::Log::Error,
+						     "Mission land item could not be read");
+				}
```
Comment (JonasPerolini) [inline src/modules/navigator/rtl.cpp:504]: Hey @bkueng, 

I'm not sure how to handle the case where the land item cache is still pending. In that case it's not a read error so sending the critical warning can be misleading. 

On the other hand, re-evaluating `setRtlTypeAndDestination` does not seem right because we've already fell back to home / safepoint. 

Any thoughts on how to handle this corner case?

Comment (JonasPerolini) [reply src/modules/navigator/rtl.cpp:504]: My preferred option is to remove: `if (!land_item_pending) {` and have pending equivalent to SD card read failure. This is the simplest option and in practice the land item should not be pending since `on_inactive()` we call `updateDatamanCache`.

Comment (JonasPerolini) [reply src/modules/navigator/rtl.cpp:504]: removed `if (!land_item_pending) {` here https://github.com/PX4/PX4-Autopilot/pull/27707/changes/38ccd0dcf46525de3fd96bf99b3b69a94713a153

**Comment by bkueng** [comment]
I'll try to find some time for a review.

**Comment by JonasPerolini** [comment]
> _I'll try to find some time for a review._

Thanks @bkueng, please note that I have created a simpler PR as a first step: https://github.com/PX4/PX4-Autopilot/pull/27804. That PR does not contain the mission route cache (which is the trickier part of the PR)

**Comment by JonasPerolini** [comment]
Hey @bkueng, thanks for merging https://github.com/PX4/PX4-Autopilot/pull/27804. I've rebased this PR on top of it and also addressed your comment regarding copy efficiency of `missionLandItemCacheFullyLoaded()` https://github.com/PX4/PX4-Autopilot/pull/27804#discussion_r3588045727 (see details below).

Summary of the changes: 

Adds a full mission route cache to Navigator, next to the existing safe point and land item caches.

It keeps the whole uploaded mission in RAM so route planning can scan the geometry without blocking on dataman or the SD card. Size comes from `CONFIG_NAVIGATOR_FULL_MISSION_CACHE_SIZE` (56 bytes per item, allocated once a t startup). Default is 0 on NuttX, so the code is compiled out and boards pay nothing unless they opt in.

It fills asynchronously while armed, one item per navigator cycle, and with short bounded blocking reads while disarmed so it is warm before takeoff. Missions bigger than the configured size are reported and stay uncached.

As suggested in https://github.com/PX4/PX4-Autopilot/pull/27804#discussion_r3588045727 , the planner borrows the items through a `MissionView` instead of copying, with a generation counter so a stale view is detectable. `syncMissionItem()` keeps the cache in sync when a mission item is written back.

No consumer yet, that comes with the route planner PR on top of this one. 

Unit tests cover loading, reload on source change, retries, oversized missions, patching and the blocking path.

---

Please note that the mission cache is not a `DatamanCache`, it now stores `mission_item_s` directly

`DatamanCache` keeps whole dataman responses in its slots, so a slot is 80 bytes
against 56 for a bare `mission_item_s`. At 500 items that is 40 kB instead of
28 kB. Lookups are also a linear scan over all slots, and queuing N loads scans
N times, so filling the cache is O(N^2). That is fine for the 10 slot sliding
cache in `MissionBase`, less so for a full route.

Storing `mission_item_s` in a flat array gives:

- 56 bytes per item, sized once at startup from `CONFIG_RTL_MISSION_CACHE_SIZE`
- O(1) indexed access instead of a scan
- `getMissionView()`, which hands the planner a `const mission_item_s*` over the
  whole route with no copying

The tradeoff is that this cache drives its own `DatamanClient` and load state machine instead of reusing the one from `DatamanCache`, so the request, retries and stale response handling live here. In exchange it consumes a response and issues the next read in the same `update()` call, where `DatamanCache` needs two calls per item, so warm-up is twice as fast.

File: src/modules/navigator/mission_route_cache.h [hunk 2]
```diff
@@ -174,12 +265,23 @@ class MissionRouteCache : public mission_route::Provider
...
-	mutable DatamanCache _dataman_cache_safepoint{"navigator_dm_cache_route_safepoint", kInitialSafePointCacheSize};
+	// Navigator accesses MissionRouteCache from one serialized task loop. The
+	// typed mission view can therefore be borrowed for one synchronous planning call.
+#if CONFIG_NAVIGATOR_FULL_MISSION_CACHE_SIZE > 0
+	mission_item_s *_mission_items {nullptr};
+	DatamanClient _dataman_client_mission{};
+	orb_advert_t *_mavlink_log_pub{nullptr};
+	// Keep outside of MissionCacheState: it must survive `_mission = {}` reset
+	uint32_t _mission_generation{0};
+	MissionRequest _mission_request{};
```
Comment (bkueng) [inline src/modules/navigator/mission_route_cache.h]: Not using `DatamanCache` here makes sense.
I have not looked at the details, but I'm wondering if it would make the ifdef's simpler if you moved this and the relevant implementation into a separate class like FullMissionCache?

**Review by bkueng — COMMENTED** [review]
High-level, instead of the blocking reads, you could add the dataman client's subscription fd to the navigator's fd poll list. Then it would always get notified immediately and the load time would also be better while armed (without the need for separate logic).

**Comment by JonasPerolini** [comment]
Thanks for the review @bkueng, I agree on both counts. 

Code updated: 
1. Extracted all optional full-route state and logic into full_mission_cache.cpp / h --> simplifies the if def. `full_mission_cache`:
   - Owns the mission buffer, Dataman client, generation/request state, retries, synchronization, and capacity reporting.
   - `MissionRouteCache` preserves its existing public types/API.
   - Cache-size-zero builds instantiate neither a Dataman client nor a mission buffer.
2. Added a fourth Navigator poll entry --> not constrained by the navigator rate anymore.
   - The fd is enabled only while a full-mission read is pending. This is important because `dataman_response` is shared; polling it while idle could cause `Navigator` to spin on unrelated responses.
   - A response is still copied and validated by `DatamanClient::update()`. Poll ready does not mean successful completion.
   - The next item is queued immediately after consuming the previous response.
3. Removed the disarmed synchronous read, the time budgets, and related Kconfig comments. Loading is now fully asynchronous and identical while armed or disarmed.
4. Added perf counters for the full mission route cache load
   - To test: `perf reset`, Upload or change the mission, `perf`
5. Extend unit tests coverage for polling, retries, source replacement, perf timing, and  cache synchronization.
6. `DO_JUMP` state `coherent:Successful` counter increments and resets are updated into the route cache. Keeps the cached counters synchronized, so route planning uses the correct number of remaining loops

File: src/modules/navigator/mission_route_cache.h (hunk 2 above)
Comment (JonasPerolini) [reply src/modules/navigator/mission_route_cache.h]: Agreed, move into `full_mission_cache`

File: src/modules/navigator/navigator_main.cpp [hunk 3]
```diff
@@ -257,6 +259,9 @@ void Navigator::run()
 
 	while (!should_exit()) {
 
+		// Poll Dataman only while the full-mission cache has a pending read.
+		fds[3].fd = _mission_route_cache.fullMissionResponseSubscription();
```
Comment (bkueng) [inline src/modules/navigator/navigator_main.cpp:265]: The poll could now return at a high rate during an update. Most updates are guarded by topic updates, but the [run](https://github.com/JonasPerolini/PX4-Autopilot/blob/2ba2fa16ae7750c59da65cf6bdd67537d2842dd4/src/modules/navigator/navigator_main.cpp#L1065) call is not.
I think it's worth limiting that update rate, maybe in sync with local position updates in combination with a minimum update rate.

File: src/modules/navigator/full_mission_cache.cpp [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+		const mission_item_s &mission_item)
+{
+	if (!_mission.source_valid || _mission.too_large
+	    || !missionMatchesCache(mission) || index < 0 || index >= _mission.count) {
+		return SyncResult::kRejected;
+	}
+
+	if (_mission.ready) {
+		_mission_items[index] = mission_item;
+		advanceMissionGeneration();
```
Comment (bkueng) [inline src/modules/navigator/full_mission_cache.cpp:174]: Does this actually need to discard as well?

**Review by bkueng — COMMENTED** [review]
Nice work!
I did not check the logic in full detail, but overall it looks really good. Good to see you added quite a few tests to cover corner cases.
Just fyi I won't be around for the next 2 weeks.

File: src/modules/navigator/navigator_main.cpp (hunk 3 above)
Comment (JonasPerolini) [reply src/modules/navigator/navigator_main.cpp:265]: Agreed. Updated the code such that Dataman completions still advance the cache immediately, but full Navigator processing now runs only on its normal input updates or when the 50 ms minimum-rate fallback is due. This limits Navigator mode updates to the existing 20 Hz local position update.

 If you think that the Dataman only operation also needs a min rate, please let me know.

File: src/modules/navigator/full_mission_cache.cpp (hunk 4 above)
Comment (JonasPerolini) [reply src/modules/navigator/full_mission_cache.cpp:174]: The cache is patched only after the Dataman write succeeds, and advancing its generation invalidates previously borrowed views. The complete cache therefore stays available without reloading the whole mission.

Discard: 
1. The old borrowed view becomes stale once the generation is advanced. --> yes
2. pending response: no need, `_mission.ready` means there are no pending responses. --> no
3. cache ready: no need, here's the flow:
   a. Dataman write succeeds
   b. syncMissionItem()
   c. patch cached item
   d. advance generation
   e. old views become stale
   f. new views immediately see the updated item

**Comment by JonasPerolini** [comment]
Thanks for the review @bkueng and acknowledged for the 2 weeks. The code is ready on my side.

Followed your suggestion and updated the code to skip full navigator processing based on dataman poll. Navigator advances as it used to. Dataman completions only advance the cache immediately. If you think that the Dataman only operation also needs a min rate, please let me know.

Updated the PR's description to match our latest changes (and the ones from the previous PR). 

Documented the mission route cache https://github.com/PX4/PX4-Autopilot/pull/27707/changes/e161b2e5daee4144829e66a04701d61847ffa029 keeping it compatible with `docs/en/concept/mission_route_planning.md` that will come in the next PR https://github.com/PX4/PX4-Autopilot/pull/27687 (FYI @hamishwillee).

File: src/modules/navigator/navigator_main.cpp
Comment (bkueng) [inline src/modules/navigator/navigator_main.cpp:306]: Shouldn't this be:
```suggestion
		if (mission_received || (fds[3].revents & POLLIN) || run_navigator_update) {
```

File: src/modules/navigator/full_mission_cache.cpp (hunk 4 above)
Comment (bkueng) [reply src/modules/navigator/full_mission_cache.cpp:174]: Ah ok I thought the generation also triggered a full cache invalidation. That makes sense then.

File: src/modules/navigator/navigator_main.cpp (hunk 3 above)
Comment (bkueng) [reply src/modules/navigator/navigator_main.cpp:265]: Yes for handling timeouts it should be better to also run it with `run_navigator_update`.

File: src/modules/navigator/full_mission_cache.cpp (hunk 4 above)
Comment (JonasPerolini) [reply src/modules/navigator/full_mission_cache.cpp:174]: Ok, closing this thread. Just a heads up that I've added some docs to clarify how the mission route cache works.

File: src/modules/navigator/navigator_main.cpp
Comment (JonasPerolini) [reply src/modules/navigator/navigator_main.cpp:306]: Yes @bkueng  and it is already indirectly the case so not code changes are needed.

`mission_received` remains true once we've received a mission (so during the whole cache update process), it is never reset to false (only init as false). Therefore, we will call `_mission_route_cache.update(mission);` on `run_navigator_update`

Having the `|| run_navigator_update` can lead to calling `_mission_route_cache.update(mission);` with an un-init mission at first which would contain `mission_id 0`.

Also confirmed that mission deletion is properly handled: the mavlink `MISSION_CLEAR_ALL` just publishes a new empty mission so no need to reset `mission_received` to false since we still have a mission (just empty). The cache is properly invalidated.

File: src/modules/navigator/navigator_main.cpp (hunk 3 above)
Comment (JonasPerolini) [reply src/modules/navigator/navigator_main.cpp:265]: Checked again and we actually already run at `run_navigator_update` because `mission_received` remains true once we've received a mission, so on `run_navigator_update` we will call `_mission_route_cache.update(mission);` and thus handle timeouts / backoffs properly, sorry for the confusion.

see https://github.com/PX4/PX4-Autopilot/pull/27707#discussion_r3774442481 for more details.

File: src/modules/navigator/navigator_main.cpp
Comment (bkueng) [reply src/modules/navigator/navigator_main.cpp:306]: I should have checked the context. I thought `mission_received` was declared within the loop. This is all good then.

**Review by bkueng — APPROVED** [review]

File: src/modules/navigator/rtl.cpp (hunk 1 above)
Comment (JonasPerolini) [reply src/modules/navigator/rtl.cpp:504]: resolving thread

**Comment by JonasPerolini** [comment]
Thank you for the feedback @bkueng, merged the code, moving on to the next PR that will use the mission route cache: https://github.com/PX4/PX4-Autopilot/pull/27687

**Comment by hamishwillee** [comment]
Congratulations on getting this in!
